DDF-1555: Implement solution for cardinality issues that cause unpredictable configuration

lessarderic opened the pull request:
- Added a configuration admin listener that reads and writes custom configuration
   files that properly support cardinality. The current implementation uses
   Felix' ConfigurationHandler class to read and write the files.
 - Added a file watcher to configuration directory. Previously, there
   wasn't a way to get changes that were made directly in the
   filesystem as opposed to the given administration ui or commands.
   With this change, ddf will pick up any changes to config files that are
   made (created, modified, deleted) and propagate them to the Config Admin.
 - Listeners added as a feature that installs at Karaf start up.
 - Added ConfigurationPropertiesComparator class in platform-util to compare
   OSGi Configuration properties and removed equal() method from FileHandlerImpl
   class.
 - Removed cardinality workaround in policy manager.

Hero:
@brendan-hofmann

Reviewers:
@pklinef 
@figliold 
@andrewkfiedler 
@garrettfreibott 
@oconnormi 
@clockard 

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/codice/ddf/261)
<!-- Reviewable:end -->

brendan-hofmann commented:
`Tests in error: 
  testConfiguration(org.codice.security.policy.context.PolicyManagerTest): java.lang.String cannot be cast to [Ljava.lang.String;
  testSetPropertiesIgnoresNullMap(org.codice.security.policy.context.PolicyManagerTest): java.lang.String cannot be cast to [Ljava.lang.String;`
<!-- Reviewable comment -K0hH-0VX1gSHJ1BF3Gt:-1166468624 -->

---

Reviewed 29 of 29 files at r1.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

<sup>**[platform/platform-app/src/main/resources/features.xml, line 531 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/261#-K0h8Pt2Vd6nyrx_pln-)** ([raw file](https://github.com/codice/ddf/blob/06e422b5e591769f8734d609a85def874926102a/platform/platform-app/src/main/resources/features.xml#L531)):</sup>
Shouldn't this be installed by default?
<!-- Reviewable comment -K0h8Pt2Vd6nyrx_pln0:1329563625 -->

---

<sup>**[platform/platform-configuration-listener/src/main/java/org/codice/ddf/configuration/listener/ConfigurationAdminListener.java, line 73 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/261#-K0hB4ZlBwMeL4flDNvm)** ([raw file](https://github.com/codice/ddf/blob/06e422b5e591769f8734d609a85def874926102a/platform/platform-configuration-listener/src/main/java/org/codice/ddf/configuration/listener/ConfigurationAdminListener.java#L73)):</sup>
This doesn't make sense to me at info. Debug maybe? Unless it's potentially a problem, in which case it should be warn.
<!-- Reviewable comment -K0hB4ZmRQfKGZOkBHWX:104838302 -->

---

<sup>**[platform/platform-configuration-listener/src/main/java/org/codice/ddf/configuration/listener/ConfigurationAdminListener.java, line 83 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/261#-K0hBL_8OG3u8tPkRIGJ)** ([raw file](https://github.com/codice/ddf/blob/06e422b5e591769f8734d609a85def874926102a/platform/platform-configuration-listener/src/main/java/org/codice/ddf/configuration/listener/ConfigurationAdminListener.java#L83)):</sup>
"configuration information of configadmin"
that phrase doesn't make sense. Typo?
<!-- Reviewable comment -K0hBL_8OG3u8tPkRIGK:1194001242 -->

---

<sup>**[platform/platform-configuration-listener/src/main/java/org/codice/ddf/configuration/store/ConfigurationFilesPoller.java, line 102 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/261#-K0hDXx5_iWc2ojk3rOd)** ([raw file](https://github.com/codice/ddf/blob/06e422b5e591769f8734d609a85def874926102a/platform/platform-configuration-listener/src/main/java/org/codice/ddf/configuration/store/ConfigurationFilesPoller.java#L102)):</sup>
Switching on raw strings? Why aren't we using constants for `kind`?
<!-- Reviewable comment -K0hDXx5_iWc2ojk3rOe:1354259709 -->

---

<sup>**[platform/platform-configuration-listener/src/main/java/org/codice/ddf/configuration/store/FileHandler.java, line 50 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/261#-K0hDsSEoqvuBtP4Svql)** ([raw file](https://github.com/codice/ddf/blob/06e422b5e591769f8734d609a85def874926102a/platform/platform-configuration-listener/src/main/java/org/codice/ddf/configuration/store/FileHandler.java#L50)):</sup>
"If not configuration existed"? Typo?
<!-- Reviewable comment -K0hDsSFeG0JlLCHTfW_:-2089171183 -->

---

<sup>**[platform/platform-configuration-listener/src/main/java/org/codice/ddf/configuration/store/FileHandler.java, line 71 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/261#-K0hE6JoFVnPwcC1fgf0)** ([raw file](https://github.com/codice/ddf/blob/06e422b5e591769f8734d609a85def874926102a/platform/platform-configuration-listener/src/main/java/org/codice/ddf/configuration/store/FileHandler.java#L71)):</sup>
Only one listener per pid/file, or only one listener period?

Either way, this is a weird pattern to me. Why not just provide a way to unregister listeners?
<!-- Reviewable comment -K0hE6JpC0LsJYM-NQO7:2020926473 -->

---

<sup>**[platform/util/pom.xml, line 46 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/261#-K0hIQYq8Uk1K5Kue1Wx)** ([raw file](https://github.com/codice/ddf/blob/06e422b5e591769f8734d609a85def874926102a/platform/util/pom.xml#L46)):</sup>
Why not use the maven property for this version?

I'm not going to comment on every version in every pom, but in general, same question. I'd rather see properties/managed dependencies. 
<!-- Reviewable comment -K0hIQYreyXBTJ-irKWJ:1844843691 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/261)
<!-- Sent from Reviewable.io -->

lessarderic commented:
Should be fixed now.
<!-- Reviewable comment -K0hN3xuKNC7pU3tFIFa:1607712313 -->

---

Review status: 27 of 30 files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

<sup>**[platform/platform-app/src/main/resources/features.xml, line 531 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/261#-K0h8Pt2Vd6nyrx_pln-)** ([raw file](https://github.com/codice/ddf/blob/06e422b5e591769f8734d609a85def874926102a/platform/platform-app/src/main/resources/features.xml#L531)):</sup>
This attribute only applies when the feature is dropped in the deploy folder:
>Installation mode. Can be either manual or auto. Specifies whether the feature should be automatically installed when dropped inside the deploy folder. Note: This attribute doesn't affect feature descriptors that are installed from the command line or as part of the org.apache.karaf.features.cfg.
<!-- Reviewable comment -K0hSw0fewnL9-jkXWMf:1468719954 -->

---

<sup>**[platform/platform-configuration-listener/src/main/java/org/codice/ddf/configuration/listener/ConfigurationAdminListener.java, line 73 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/261#-K0hB4ZlBwMeL4flDNvm)** ([raw file](https://github.com/codice/ddf/blob/06e422b5e591769f8734d609a85def874926102a/platform/platform-configuration-listener/src/main/java/org/codice/ddf/configuration/listener/ConfigurationAdminListener.java#L73)):</sup>
Changed.
<!-- Reviewable comment -K0hO2AlNNsA8-jERVSX:-609445857 -->

---

<sup>**[platform/platform-configuration-listener/src/main/java/org/codice/ddf/configuration/listener/ConfigurationAdminListener.java, line 83 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/261#-K0hBL_8OG3u8tPkRIGJ)** ([raw file](https://github.com/codice/ddf/blob/06e422b5e591769f8734d609a85def874926102a/platform/platform-configuration-listener/src/main/java/org/codice/ddf/configuration/listener/ConfigurationAdminListener.java#L83)):</sup>
Changed to `Returns the configuration object associated with a specific PID and configuration event.`
<!-- Reviewable comment -K0hONXpklGreRoZaglZ:-294189836 -->

---

<sup>**[platform/platform-configuration-listener/src/main/java/org/codice/ddf/configuration/store/ConfigurationFilesPoller.java, line 102 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/261#-K0hDXx5_iWc2ojk3rOd)** ([raw file](https://github.com/codice/ddf/blob/06e422b5e591769f8734d609a85def874926102a/platform/platform-configuration-listener/src/main/java/org/codice/ddf/configuration/store/ConfigurationFilesPoller.java#L102)):</sup>
Changed to use `StandardWatchEventKinds` instead of hard-coded strings. Had to switch to `if/else` statements since `switch` only supports constant expressions.
<!-- Reviewable comment -K0hTGtgUMm3h3upifJY:111426577 -->

---

<sup>**[platform/platform-configuration-listener/src/main/java/org/codice/ddf/configuration/store/FileHandler.java, line 50 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/261#-K0hDsSEoqvuBtP4Svql)** ([raw file](https://github.com/codice/ddf/blob/06e422b5e591769f8734d609a85def874926102a/platform/platform-configuration-listener/src/main/java/org/codice/ddf/configuration/store/FileHandler.java#L50)):</sup>
Reworded the comment.
<!-- Reviewable comment -K0hV6FgQxAgxF3ppljI:544429309 -->

---

<sup>**[platform/platform-configuration-listener/src/main/java/org/codice/ddf/configuration/store/FileHandler.java, line 71 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/261#-K0hE6JoFVnPwcC1fgf0)** ([raw file](https://github.com/codice/ddf/blob/06e422b5e591769f8734d609a85def874926102a/platform/platform-configuration-listener/src/main/java/org/codice/ddf/configuration/store/FileHandler.java#L71)):</sup>
No use case right now to support unregistration, but we could always accept `null` to allow for that. Thoughts?
<!-- Reviewable comment -K0hVXKd2OGAMb9tylkF:-187271336 -->

---

<sup>**[platform/util/pom.xml, line 46 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/261#-K0hIQYq8Uk1K5Kue1Wx)** ([raw file](https://github.com/codice/ddf/blob/06e422b5e591769f8734d609a85def874926102a/platform/util/pom.xml#L46)):</sup>
No property defined for this anywhere so I removed the `<version>` element for now like we did elsewhere.
<!-- Reviewable comment -K0hXxKtNw4ApqMLAC1V:2017620793 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/261)
<!-- Sent from Reviewable.io -->

brendan-hofmann commented:
Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

<sup>**[platform/platform-app/src/main/resources/features.xml, line 531 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/261#-K0h8Pt2Vd6nyrx_pln-)** ([raw file](https://github.com/codice/ddf/blob/06e422b5e591769f8734d609a85def874926102a/platform/platform-app/src/main/resources/features.xml#L531)):</sup>
Understood.

I still feel that, since it fulfills such a fundamental purpose, it should be part of the platform app. However, perhaps I am mistaken in that and/or am only inclined towards our convention out of habit. Thoughts?
<!-- Reviewable comment -K0hZ0RfKC5n2l9mDNyg:-479419119 -->

---

<sup>**[platform/platform-configuration-listener/src/main/java/org/codice/ddf/configuration/store/ConfigurationFilesPoller.java, line 102 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/261#-K0hDXx5_iWc2ojk3rOd)** ([raw file](https://github.com/codice/ddf/blob/06e422b5e591769f8734d609a85def874926102a/platform/platform-configuration-listener/src/main/java/org/codice/ddf/configuration/store/ConfigurationFilesPoller.java#L102)):</sup>
Yeah, it's kinda weird how they did that. Still better than raw strings, imho, so thanks for changing.
<!-- Reviewable comment -K0hZbawjaVJQSMmVjDE:529684640 -->

---

<sup>**[platform/platform-configuration-listener/src/main/java/org/codice/ddf/configuration/store/FileHandler.java, line 71 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/261#-K0hE6JoFVnPwcC1fgf0)** ([raw file](https://github.com/codice/ddf/blob/06e422b5e591769f8734d609a85def874926102a/platform/platform-configuration-listener/src/main/java/org/codice/ddf/configuration/store/FileHandler.java#L71)):</sup>
I was more questioning the single listener aspect. If you allow multiple listeners then you also need more granular unregistering. So the question is really why limit it to a single listener? Looking at the ChangeListener interface, you could register different ChangeListeners for different pids or sets of pids.

Maybe I'm overthinking this, but the implications of the two apis seem inconsistent. You could achieve the same behavior with one monolithic listener, I guess.
<!-- Reviewable comment -K0ha41mTFxhTASQjP6D:803703439 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/261)
<!-- Sent from Reviewable.io -->

clockard commented on platform/security/policy/security-policy-context/src/main/java/org/codice/ddf/security/policy/context/impl/PolicyManager.java:
```diff
@@ -481,24 +442,10 @@ public void setRealms(List<String> realms) {
-            contexts = PropertyResolver.resolveProperties(contexts);
-            String[] contextsArr = contexts.split(",");
-            whiteListContexts = Arrays.asList(contextsArr);
+            whiteListContexts = contexts;
```
This will not work with the new variable replacement that was added. WhiteListContexts strings can contain variables that must be replaced on ingest. Looks like with the way it is written right now we need to loop through the list and replace each value with output from  running PropertyResolver.resolveProperties on each item.

lessarderic commented:
Review status: 29 of 30 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

<sup>**[platform/platform-configuration-listener/src/main/java/org/codice/ddf/configuration/store/FileHandler.java, line 71 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/261#-K0hE6JoFVnPwcC1fgf0)** ([raw file](https://github.com/codice/ddf/blob/06e422b5e591769f8734d609a85def874926102a/platform/platform-configuration-listener/src/main/java/org/codice/ddf/configuration/store/FileHandler.java#L71)):</sup>
These are really meant to be internal interfaces at this point and we only expect one listener to register at startup and never remove itself. Since this is our only use case right now, I wanted to keep the interfaces simple. We can always revisit them later if we think there could be more than one listener, or listeners that may be interested in specific PIDs. Does that sound reasonable?
<!-- Reviewable comment -K0hgUGDxMnMIQy1B8cV:-1921499643 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/261)
<!-- Sent from Reviewable.io -->

brendan-hofmann commented:
"Addressed code review comments"
This isn't a very meaningful commit message. If the changes are really that minor, they could probably just be rebased into the existing commits.
<!-- Reviewable comment -K0hn5BFeB4J7i895QBZ:678392684 -->

---

Reviewed 7 of 7 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

<sup>**[platform/platform-configuration-listener/src/main/java/org/codice/ddf/configuration/store/FileHandler.java, line 71 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/261#-K0hE6JoFVnPwcC1fgf0)** ([raw file](https://github.com/codice/ddf/blob/06e422b5e591769f8734d609a85def874926102a/platform/platform-configuration-listener/src/main/java/org/codice/ddf/configuration/store/FileHandler.java#L71)):</sup>
Yeah, I think it's fine for now, especially since it's all documented. Just wanted to make sure we weren't sabotaging ourselves for the future. ;)
<!-- Reviewable comment -K0hnOcWPRN3F9LZscom:2063858181 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/261)
<!-- Sent from Reviewable.io -->

lessarderic commented:
<sup>**[platform/security/policy/security-policy-context/src/main/java/org/codice/ddf/security/policy/context/impl/PolicyManager.java, line 448 \[r2\]](https://reviewable.io:443/reviews/codice/ddf/261#-K0gv3fyfM0GwyFSES8b-r2-448)** ([raw file](https://github.com/codice/ddf/blob/d3ae298524394df6b2305f7e4085a7fe674410b3/platform/security/policy/security-policy-context/src/main/java/org/codice/ddf/security/policy/context/impl/PolicyManager.java#L448)):</sup>
Fixed
<!-- Reviewable comment -K0i2Y_IYJTiYmYYb7k6:-497468049 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/261)
<!-- Sent from Reviewable.io -->

lessarderic commented:
Review status: 25 of 30 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

<sup>**[platform/platform-app/src/main/resources/features.xml, line 531 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/261#-K0h8Pt2Vd6nyrx_pln-)** ([raw file](https://github.com/codice/ddf/blob/06e422b5e591769f8734d609a85def874926102a/platform/platform-app/src/main/resources/features.xml#L531)):</sup>
Since this will be used to manage all configurations, including the platform ones, we needed to make sure it would be the first bundle to come up which is why we've put it in a separate feature and put it before platform-app.

Hope that makes sense.
<!-- Reviewable comment -K0i841TibLPORQnkFLn:-792444237 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/261)
<!-- Sent from Reviewable.io -->

lessarderic commented:
Squashed commits.
<!-- Reviewable comment -K0im4R3QD6vGoFSvmbI:29772495 -->

---

Review status: 25 of 30 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/261)
<!-- Sent from Reviewable.io -->

roelens8 commented on platform/platform-configuration-listener/src/main/java/org/codice/ddf/configuration/listener/ConfigurationFileListener.java:
```diff
@@ -0,0 +1,108 @@
+    public void update(String filePid, ChangeType changeType) {
+        try {
+            switch (changeType) {
+            case CREATED:
```
There is no separate logic for the "created" case,  or is the logic supposed to be shared between "created" and "updated" ? E.g if (changeType.equals(CREATE) || changeType.equals(UPDATED)) then updateConfig(..)

andrewkfiedler replied:
The logic is meant to be shared.  Should we add a comment so the fall through on the switch case doesn't look like an accident?

brendan-hofmann commented:
Reviewed 1 of 29 files at r1, 2 of 8 files at r4, 3 of 6 files at r5, 4 of 4 files at r6, 1 of 1 files at r7, 2 of 2 files at r8.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

<sup>**[distribution/docs/src/main/resources/ManContents.adoc, line 1172 \[r8\]](https://reviewable.io:443/reviews/codice/ddf/261#-K0mPGKBZKEovA7tBRXr)** ([raw file](https://github.com/codice/ddf/blob/49b2dc39a9fb5172ad8716e58f26af5ae1131fd6/distribution/docs/src/main/resources/ManContents.adoc#L1172)):</sup>
What defines the set of escaped characters here? Can we document that?
<!-- Reviewable comment -K0mPGKBZKEovA7tBRXs:1044382728 -->

---

<sup>**[libs/common-system/pom.xml, line 34 \[r8\]](https://reviewable.io:443/reviews/codice/ddf/261#-K0mPT3WrgGtGuVhnvmH)** ([raw file](https://github.com/codice/ddf/blob/49b2dc39a9fb5172ad8716e58f26af5ae1131fd6/libs/common-system/pom.xml#L34)):</sup>
I think it's about time hamcrest was dependency-managed. ;)
<!-- Reviewable comment -K0mPT3WrgGtGuVhnvmI:1832428400 -->

---

<sup>**[platform/platform-app/src/main/resources/features.xml, line 531 \[r1\]](https://reviewable.io:443/reviews/codice/ddf/261#-K0h8Pt2Vd6nyrx_pln-)** ([raw file](https://github.com/codice/ddf/blob/06e422b5e591769f8734d609a85def874926102a/platform/platform-app/src/main/resources/features.xml#L531)):</sup>
Yep, I see now. :+1: 
<!-- Reviewable comment -K0mOxDAz_mzD-EcmgLW:1073165877 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/261)
<!-- Sent from Reviewable.io -->

lessarderic commented on distribution/docs/src/main/resources/ManContents.adoc:
```diff
@@ -1151,6 +1151,38 @@ Be sure to note the port number that replaced 8181 then enter that number in the
+The {branding} can also be configured with configuration (.config) files.
+
+[NOTE]
+====
```
@oconnormi: This is probably where we should say that only security related things can be configured using the new configuration file format. Otherwise people may try to use the new format for bundles that haven't been changed yet.

pklinef commented on platform/platform-app/src/main/resources/features.xml:
```diff
@@ -528,6 +528,11 @@
     </feature>
     <!-- End CXF Features -->
 
+    <feature name ="platform-configuration-listener" install="manual" version="${project.version}"
```
Is there a reason why this was not just added to the platform-app feature instead of adding it to org.apache.karaf.features.cfg?

pklinef commented on platform/platform-configuration-listener/pom.xml:
```diff
@@ -0,0 +1,144 @@
+        <dependency>
+            <groupId>com.google.guava</groupId>
+            <artifactId>guava</artifactId>
+            <version>17.0</version>
```
Why not 18 since this is new?

pklinef commented on platform/platform-configuration-listener/pom.xml:
```diff
@@ -0,0 +1,144 @@
+        <dependency>
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
+            <version>4.11</version>
```
Should be defined by dependency management.

pklinef commented on platform/platform-configuration-listener/pom.xml:
```diff
@@ -0,0 +1,144 @@
+        <dependency>
+            <groupId>org.hamcrest</groupId>
+            <artifactId>hamcrest-all</artifactId>
+            <version>1.3</version>
```
Should be defined by dependency management.

pklinef commented on platform/platform-configuration-listener/src/main/java/org/codice/ddf/configuration/listener/ConfigurationAdminListener.java:
```diff
@@ -0,0 +1,91 @@
+     */
+    @Override
+    public void configurationEvent(ConfigurationEvent event) {
+        Validate.notNull(event, "ConfigurationEvent is null.");
```
What happens if a ConfigurationListener throws a runtime exception?

pklinef commented on platform/security/policy/security-policy-context/src/main/resources/OSGI-INF/blueprint/blueprint.xml:
```diff
@@ -21,14 +21,40 @@
+                <value>${org.codice.ddf.system.rootContext}/SecurityTokenService</value>
+                <value>${org.codice.ddf.system.rootContext}/internal</value>
+                <value>/proxy</value>
+            </array>
```
Double check the values in master.  I know I added a SAML one related to IDP.

pklinef commented on platform/security/policy/security-policy-context/src/main/resources/OSGI-INF/blueprint/blueprint.xml:
```diff
@@ -21,14 +21,40 @@
-        <property name="whiteListContexts"
-                  value="${org.codice.ddf.system.rootContext}/SecurityTokenService,${org.codice.ddf.system.rootContext}/internal,/proxy"/>
+        <property name="authenticationTypes">
+            <array value-type="java.lang.String">
```
At some point we need to re-evaluate our use of positive cardinality which uses primitive arrays compared to using negative cardinality which will inject vectors.

pklinef commented on platform/platform-configuration-listener/src/main/resources/OSGI-INF/blueprint/blueprint.xml:
```diff
@@ -0,0 +1,91 @@
+    </bean>
+
+    <bean id="persistenceStrategy"
+          class="org.codice.ddf.configuration.store.FelixPersistenceStrategy">
```
Is it possible to support multiple persistence strategies?  Do you just have to create multiple file handlers?  I am wondering if it would make sense to have them be OSGi services and they would be injected as a reference list instead.

lessarderic commented:
Review status: 32 of 34 files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

<sup>**[distribution/docs/src/main/resources/ManContents.adoc, line 1172 \[r8\]](https://reviewable.io:443/reviews/codice/ddf/261#-K0mPGKBZKEovA7tBRXr)** ([raw file](https://github.com/codice/ddf/blob/49b2dc39a9fb5172ad8716e58f26af5ae1131fd6/distribution/docs/src/main/resources/ManContents.adoc#L1172)):</sup>
Checking to see if we really need to escape those characters. Will update the doc accordingly.
<!-- Reviewable comment -K0n5vTpxJwNSSj3eSCJ:-1432879811 -->

---

<sup>**[libs/common-system/pom.xml, line 34 \[r8\]](https://reviewable.io:443/reviews/codice/ddf/261#-K0mPT3WrgGtGuVhnvmH)** ([raw file](https://github.com/codice/ddf/blob/49b2dc39a9fb5172ad8716e58f26af5ae1131fd6/libs/common-system/pom.xml#L34)):</sup>
We already have a separate ticket for that.
<!-- Reviewable comment -K0n6bjiE504UFAZsSi4:-1040707465 -->

---

<sup>**[platform/platform-app/src/main/resources/features.xml, line 531 \[r9\]](https://reviewable.io:443/reviews/codice/ddf/261#-K0gv3fyfM0GwyFSES8_-r9-531)** ([raw file](https://github.com/codice/ddf/blob/aedbcd9efddac6b26d0d76a1214288d2a84b3f11/platform/platform-app/src/main/resources/features.xml#L531)):</sup>
As we mentioned to @brendan-hofmann (see below), we need to make sure this feature comes up before any other one so it felt safer and more explicit to put it as a separate, top-level feature. Could be convinced otherwise though. I'm not the Karaf/OSGi expert :-)
<!-- Reviewable comment -K0n6hzUTNzaGsrMHg1J:51463544 -->

---

<sup>**[platform/platform-configuration-listener/pom.xml, line 56 \[r9\]](https://reviewable.io:443/reviews/codice/ddf/261#-K0gv3fyfM0GwyFSES8i-r9-56)** ([raw file](https://github.com/codice/ddf/blob/aedbcd9efddac6b26d0d76a1214288d2a84b3f11/platform/platform-configuration-listener/pom.xml#L56)):</sup>
Updated.
<!-- Reviewable comment -K0n7GoqWwTdbVkFHZGJ:-719267080 -->

---

<sup>**[platform/platform-configuration-listener/pom.xml, line 70 \[r9\]](https://reviewable.io:443/reviews/codice/ddf/261#-K0gv3fyfM0GwyFSES8i-r9-70)** ([raw file](https://github.com/codice/ddf/blob/aedbcd9efddac6b26d0d76a1214288d2a84b3f11/platform/platform-configuration-listener/pom.xml#L70)):</sup>
We already have a separate ticket to do this.
<!-- Reviewable comment -K0n7Jk7CrW3wBxkUU1x:-532213017 -->

---

<sup>**[platform/platform-configuration-listener/pom.xml, line 76 \[r9\]](https://reviewable.io:443/reviews/codice/ddf/261#-K0gv3fyfM0GwyFSES8i-r9-76)** ([raw file](https://github.com/codice/ddf/blob/aedbcd9efddac6b26d0d76a1214288d2a84b3f11/platform/platform-configuration-listener/pom.xml#L76)):</sup>
See previous comment :-)
<!-- Reviewable comment -K0n7T62Tjlr537M-oqW:560754238 -->

---

<sup>**[platform/platform-configuration-listener/src/main/java/org/codice/ddf/configuration/listener/ConfigurationAdminListener.java, line 55 \[r9\]](https://reviewable.io:443/reviews/codice/ddf/261#-K0gv3fyfM0GwyFSES8j-r9-55)** ([raw file](https://github.com/codice/ddf/blob/aedbcd9efddac6b26d0d76a1214288d2a84b3f11/platform/platform-configuration-listener/src/main/java/org/codice/ddf/configuration/listener/ConfigurationAdminListener.java#L55)):</sup>
Good question. Spec isn't clear so it's probably safer to simply log when we receive `null` events and add a `try`/`catch (RuntimeException)` and swallow those. I'll make that change and update the unit tests.
<!-- Reviewable comment -K0n9YT6LYQZNTJ8sw-v:695318092 -->

---

<sup>**[platform/platform-configuration-listener/src/main/resources/OSGI-INF/blueprint/blueprint.xml, line 41 \[r10\]](https://reviewable.io:443/reviews/codice/ddf/261#-K0gv3fyfM0GwyFSES8t-r10-41)** ([raw file](https://github.com/codice/ddf/blob/2d5e1560eb323d1f2a060f4e6d67ab977af562b4/platform/platform-configuration-listener/src/main/resources/OSGI-INF/blueprint/blueprint.xml#L41)):</sup>
That was the idea behind the interface, but we wanted to start simple for now and do static injection instead of going to service route. We could revisit this in the next sprint if you feel it's important.
<!-- Reviewable comment -K0nA-EQctrgXW60ZQts:-638290335 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/261)
<!-- Sent from Reviewable.io -->

lessarderic commented:
<sup>**[platform/security/policy/security-policy-context/src/main/resources/OSGI-INF/blueprint/blueprint.xml, line 56 \[r10\]](https://reviewable.io:443/reviews/codice/ddf/261#-K0gv3fyfM0GwyFSES8c-r10-56)** ([raw file](https://github.com/codice/ddf/blob/2d5e1560eb323d1f2a060f4e6d67ab977af562b4/platform/security/policy/security-policy-context/src/main/resources/OSGI-INF/blueprint/blueprint.xml#L56)):</sup>
Checked and didn't see anything new.
<!-- Reviewable comment -K0nBngIVp9s9HmeuyzG:-1427481310 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/261)
<!-- Sent from Reviewable.io -->

lessarderic commented:
<sup>**[platform/security/policy/security-policy-context/src/main/resources/OSGI-INF/blueprint/blueprint.xml, line 25 \[r10\]](https://reviewable.io:443/reviews/codice/ddf/261#-K0gv3fyfM0GwyFSES8c-r10-25)** ([raw file](https://github.com/codice/ddf/blob/2d5e1560eb323d1f2a060f4e6d67ab977af562b4/platform/security/policy/security-policy-context/src/main/resources/OSGI-INF/blueprint/blueprint.xml#L25)):</sup>
Are you saying that we should revisit our metatype.xml files and move away from arrays and use `Vector`s instead? I couldn't really figure out whether the + vs - cardinality is a Metatype standard, but Felix' `ConfigurationHandler` does support both ([] for arrays and () for vectors) so our configuration code will support whatever the Metatype defines.
<!-- Reviewable comment -K0qZfPmUUtEgHYR0He2:285587339 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/261)
<!-- Sent from Reviewable.io -->

pklinef commented on distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestPlatform.java:
```diff
@@ -0,0 +1,187 @@
+                POLLER_WAIT_TIME);
+        LOGGER.debug("Configuration deleted!");
+        /*The next line (the one updating properties) will cause the test to fail.
+          For some reason it appears to generate a CM_LOCATION_CHANGED event, which
```
Getting the configuration of a PID will rebind the location to the bundle asking for the configuration.  When the pax exam probe asks for the configuration, it will change location to the probe's location.  The listener should not fail for any of the possible events including CM_LOCATION_CHANGED.

pklinef commented:
Review status: 30 of 40 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

<sup>**[platform/platform-app/src/main/resources/features.xml, line 531 \[r9\]](https://reviewable.io:443/reviews/codice/ddf/261#-K0gv3fyfM0GwyFSES8_-r9-531)** ([raw file](https://github.com/codice/ddf/blob/aedbcd9efddac6b26d0d76a1214288d2a84b3f11/platform/platform-app/src/main/resources/features.xml#L531)):</sup>
You start it right before the platform-app feature which is the feature you would add it to.  You can have it be the first feature in the platform-app feature.  It just makes it easier for downstream custom distribution to get this change.  As a separate boot feature, they will have to manually notice this change and add the feature to their custom boot features.  It is best to avoid that if it is not necessary.
<!-- Reviewable comment -K1-up0Q9vXmorGVD9ZO:-604470789 -->

---

<sup>**[platform/platform-configuration-listener/src/main/resources/OSGI-INF/blueprint/blueprint.xml, line 41 \[r10\]](https://reviewable.io:443/reviews/codice/ddf/261#-K0gv3fyfM0GwyFSES8t-r10-41)** ([raw file](https://github.com/codice/ddf/blob/2d5e1560eb323d1f2a060f4e6d67ab977af562b4/platform/platform-configuration-listener/src/main/resources/OSGI-INF/blueprint/blueprint.xml#L41)):</sup>
It just looked like there were assumptions made in the code and interface that might make it harder to support multiple implementation currently.  I am fine with making this extensible in a future ticket.
<!-- Reviewable comment -K1-vZiTIG-h6i3TQgDw:498357107 -->

---

<sup>**[platform/security/policy/security-policy-context/src/main/resources/OSGI-INF/blueprint/blueprint.xml, line 56 \[r10\]](https://reviewable.io:443/reviews/codice/ddf/261#-K0gv3fyfM0GwyFSES8c-r10-56)** ([raw file](https://github.com/codice/ddf/blob/2d5e1560eb323d1f2a060f4e6d67ab977af562b4/platform/security/policy/security-policy-context/src/main/resources/OSGI-INF/blueprint/blueprint.xml#L56)):</sup>
https://github.com/codice/ddf/blob/master/platform/security/policy/security-policy-context/src/main/resources/OSGI-INF/metatype/metatype.xml#L39

Notice the addition of ${org.codice.ddf.system.rootContext}/saml.  Also, ${org.codice.ddf.system.rootContext}/internal is now ${org.codice.ddf.system.rootContext}/internal/metrics.
<!-- Reviewable comment -K1-w7H2gW6fsDHSLi37:2048393966 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/261)
<!-- Sent from Reviewable.io -->

oconnormi commented:
<sup>**[platform/security/policy/security-policy-context/src/main/resources/OSGI-INF/blueprint/blueprint.xml, line 56 \[r10\]](https://reviewable.io:443/reviews/codice/ddf/261#-K0gv3fyfM0GwyFSES8c-r10-56)** ([raw file](https://github.com/codice/ddf/blob/2d5e1560eb323d1f2a060f4e6d67ab977af562b4/platform/security/policy/security-policy-context/src/main/resources/OSGI-INF/blueprint/blueprint.xml#L56)):</sup>
Just added those additional contexts
<!-- Reviewable comment -K10PDXEWLsWExz7r-A6:-1233226693 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/261)
<!-- Sent from Reviewable.io -->

lessarderic commented:
Review status: 19 of 31 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

<sup>**[platform/platform-app/src/main/resources/features.xml, line 531 \[r9\]](https://reviewable.io:443/reviews/codice/ddf/261#-K0gv3fyfM0GwyFSES8_-r9-531)** ([raw file](https://github.com/codice/ddf/blob/aedbcd9efddac6b26d0d76a1214288d2a84b3f11/platform/platform-app/src/main/resources/features.xml#L531)):</sup>
Removed from `org.apache.karaf.features.cfg` and moved at the top of the `platform-app` feature. Also commented it out for now until the full implementation is ready.
<!-- Reviewable comment -K12TBxsV6nYoU5Q0X6E:-366291945 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/261)
<!-- Sent from Reviewable.io -->

lessarderic commented:
@brendan-hofmann: All the review comments have been addressed and the itests completed. Since the platform-configuration-listeners feature is now disabled by default, the only thing you should have to do is make sure both DDF and DIB still work as expected after building them. No need to test anything else. Let me know if you have any questions.

brendan-hofmann commented:
Passed a full build, test, and manual fiddle-faddle. LGTM. :+1:

lessarderic commented:
About to make major changes to the code so closing this PR for now. Will open a new one when the changes are ready to be re-reviewed.
